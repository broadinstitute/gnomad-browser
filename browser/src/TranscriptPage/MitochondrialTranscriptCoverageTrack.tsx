import React from 'react'

import {
  DatasetId,
  labelForDataset,
  referenceGenome,
  hasMitochondrialGenomeCoverage,
} from '@gnomad/dataset-metadata/metadata'
import CoverageTrack from '../CoverageTrack'
import Query from '../Query'
import StatusMessage from '../StatusMessage'

const operationName = 'MitochondrialCoverageInTranscript'
const query = `
query ${operationName}($transcriptId: String!, $datasetId: DatasetId!, $referenceGenome: ReferenceGenomeId!) {
  transcript(transcript_id: $transcriptId, reference_genome: $referenceGenome) {
    mitochondrial_coverage(dataset: $datasetId) {
      pos
      mean
      median
      over_100
      over_1000
    }
  }
}
`

type Props = {
  datasetId: DatasetId
  transcriptId: string
}

const MitochondrialTranscriptCoverageTrack = ({ datasetId, transcriptId }: Props) => {
  if (!hasMitochondrialGenomeCoverage(datasetId)) {
    return (
      <StatusMessage>
        Mitochondrial genome coverage is not available in {labelForDataset(datasetId)}
      </StatusMessage>
    )
  }

  return (
    <Query
      operationName={operationName}
      query={query}
      variables={{ transcriptId, datasetId, referenceGenome: referenceGenome(datasetId) }}
      loadingMessage="Loading coverage"
      loadingPlaceholderHeight={220}
      errorMessage="Unable to load coverage"
      success={(data: any) => data.transcript && data.transcript.mitochondrial_coverage}
    >
      {({ data }: any) => {
        const coverage = [
          {
            color: 'rgb(115, 171, 61)',
            buckets: data.transcript.mitochondrial_coverage,
            name: 'mitochondrial genome',
            opacity: 0.7,
          },
        ]

        return (
          // @ts-expect-error TS(2769) FIXME: No overload matches this call.
          <CoverageTrack
            coverageOverThresholds={[100, 1000]}
            datasets={coverage}
            filenameForExport={() => `${transcriptId}_coverage`}
            height={190}
            maxCoverage={3000}
            datasetId={datasetId}
          />
        )
      }}
    </Query>
  )
}

export default MitochondrialTranscriptCoverageTrack
