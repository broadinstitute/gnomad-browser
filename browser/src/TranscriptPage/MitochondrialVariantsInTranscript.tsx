import React from 'react'

import ClinvarVariantTrack from '../ClinvarVariantsTrack/ClinvarVariantTrack'
import formatClinvarDate from '../ClinvarVariantsTrack/formatClinvarDate'
import { labelForDataset } from '../../../dataset-metadata/metadata'
import { referenceGenomeForDataset } from '../datasets'
import Link from '../Link'
import Query from '../Query'
import filterVariantsInZoomRegion from '../RegionViewer/filterVariantsInZoomRegion'
import StatusMessage from '../StatusMessage'
import { TrackPageSection } from '../TrackPage'
import MitochondrialVariants from '../MitochondrialVariantList/MitochondrialVariants'
import annotateVariantsWithClinvar from '../VariantList/annotateVariantsWithClinvar'

const query = `
query MitochondrialVariantsInTranscript($transcriptId: String!, $datasetId: DatasetId!, $referenceGenome: ReferenceGenomeId!) {
  meta {
    clinvar_release_date
  }
  transcript(transcript_id: $transcriptId, reference_genome: $referenceGenome) {
    clinvar_variants {
      clinical_significance
      clinvar_variation_id
      gnomad {
        exome {
          ac
          an
          filters
        }
        genome {
          ac
          an
          filters
        }
      }
      gold_stars
      hgvsc
      hgvsp
      in_gnomad
      major_consequence
      pos
      review_status
      transcript_id
      variant_id
    }
    mitochondrial_variants(dataset: $datasetId) {
      ac_het
      ac_hom
      an
      consequence
      filters
      flags
      gene_id
      gene_symbol
      transcript_id
      hgvsc
      hgvsp
      lof
      lof_filter
      lof_flags
      max_heteroplasmy
      pos
      reference_genome
      variant_id
    }
  }
}
`

type OwnProps = {
  datasetId: string
  transcript: {
    transcript_id: string
  }
  zoomRegion?: {
    start: number
    stop: number
  }
}

// @ts-expect-error TS(2456) FIXME: Type alias 'Props' circularly references itself.
type Props = OwnProps & typeof MitochondrialVariantsInTranscript.defaultProps

// @ts-expect-error TS(7022) FIXME: 'MitochondrialVariantsInTranscript' implicitly has... Remove this comment to see the full error message
const MitochondrialVariantsInTranscript = ({
  datasetId,
  transcript,
  zoomRegion,
  ...rest
}: Props) => {
  if (datasetId === 'exac' || datasetId.startsWith('gnomad_r2')) {
    return (
      <StatusMessage>
        Mitochondrial variants are not available in {labelForDataset(datasetId)}
        <br />
        <br />
        <Link
          to={`/transcript/${transcript.transcript_id}?dataset=gnomad_r3`}
          preserveSelectedDataset={false}
        >
          View this transcript in gnomAD v3.1 to see mitochondrial variants
        </Link>
      </StatusMessage>
    )
  }

  return (
    <Query
      query={query}
      variables={{
        datasetId,
        transcriptId: transcript.transcript_id,
        referenceGenome: referenceGenomeForDataset(datasetId),
      }}
      loadingMessage="Loading variants"
      errorMessage="Unable to load variants"
      success={(data: any) => data.transcript && data.transcript.mitochondrial_variants}
    >
      {({ data }: any) => {
        data.transcript.mitochondrial_variants.forEach((v: any) => {
          /* eslint-disable no-param-reassign */
          if (v.an !== 0) {
            v.af = (v.ac_het + v.ac_hom) / v.an
            v.af_het = v.ac_het / v.an
            v.af_hom = v.ac_hom / v.an
          } else {
            v.af = 0
            v.af_het = 0
            v.af_hom = 0
          }
          v.hgvs = v.hgvsp || v.hgvsc
          /* eslint-enable no-param-reassign */
        })

        return (
          <>
            <TrackPageSection>
              <h2>ClinVar variants</h2>
            </TrackPageSection>
            {data.transcript.clinvar_variants.length > 0 ? (
              <>
                <ClinvarVariantTrack
                  referenceGenome={referenceGenomeForDataset(datasetId)}
                  transcripts={[transcript]}
                  variants={filterVariantsInZoomRegion(
                    data.transcript.clinvar_variants,
                    zoomRegion
                  )}
                />
                <TrackPageSection as="p">
                  Data displayed here is from ClinVar&apos;s{' '}
                  {formatClinvarDate(data.meta.clinvar_release_date)} release.
                </TrackPageSection>
              </>
            ) : (
              <TrackPageSection as="p">
                No ClinVar variants found in this transcript.
              </TrackPageSection>
            )}

            <MitochondrialVariants
              {...rest}
              clinvarReleaseDate={data.meta.clinvar_release_date}
              context={transcript}
              datasetId={datasetId}
              exportFileName={`gnomad_mitochondrial_variants_${transcript.transcript_id}`}
              variants={filterVariantsInZoomRegion(
                annotateVariantsWithClinvar(
                  data.transcript.mitochondrial_variants,
                  data.transcript.clinvar_variants
                ),
                zoomRegion
              )}
            />
          </>
        )
      }}
    </Query>
  )
}

MitochondrialVariantsInTranscript.defaultProps = {
  zoomRegion: null,
}

export default MitochondrialVariantsInTranscript
