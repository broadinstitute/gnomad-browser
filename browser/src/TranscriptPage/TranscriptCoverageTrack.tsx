import React from 'react'

import { referenceGenome } from '@gnomad/dataset-metadata/metadata'
import { coverageDataset } from '../coverage'
import { coverageConfigClassic, coverageConfigNew } from '../coverageStyles'
import CoverageTrack from '../CoverageTrack'
import Query from '../Query'

const operationName = 'TranscriptCoverage'
const coverageQuery = `
query ${operationName}($transcriptId: String!, $datasetId: DatasetId!, $referenceGenome: ReferenceGenomeId!, $includeExomeCoverage: Boolean!, $includeGenomeCoverage: Boolean!) {
  transcript(transcript_id: $transcriptId, reference_genome: $referenceGenome) {
    coverage(dataset: $datasetId) {
      exome @include(if: $includeExomeCoverage) {
        pos
        mean
        median
        over_1
        over_5
        over_10
        over_15
        over_20
        over_25
        over_30
        over_50
        over_100
      }
      genome @include(if: $includeGenomeCoverage) {
        pos
        mean
        median
        over_1
        over_5
        over_10
        over_15
        over_20
        over_25
        over_30
        over_50
        over_100
      }
    }
  }
}
`

type OwnProps = {
  datasetId: string
  transcriptId: string
  includeExomeCoverage?: boolean
  includeGenomeCoverage?: boolean
}

// @ts-expect-error TS(2456) FIXME: Type alias 'Props' circularly references itself.
type Props = OwnProps & typeof TranscriptCoverageTrack.defaultProps

// @ts-expect-error TS(7022) FIXME: 'TranscriptCoverageTrack' implicitly has type 'any... Remove this comment to see the full error message
const TranscriptCoverageTrack = ({
  datasetId,
  transcriptId,
  includeExomeCoverage,
  includeGenomeCoverage,
}: Props) => {
  return (
    <Query
      operationName={operationName}
      query={coverageQuery}
      variables={{
        transcriptId,
        datasetId: coverageDataset(datasetId),
        referenceGenome: referenceGenome(coverageDataset(datasetId)),
        includeExomeCoverage,
        includeGenomeCoverage,
      }}
      loadingMessage="Loading coverage"
      loadingPlaceholderHeight={220}
      errorMessage="Unable to load coverage"
      success={(data: any) => {
        if (!data.transcript || !data.transcript.coverage) {
          return false
        }
        const exomeCoverage = includeExomeCoverage ? data.transcript.coverage.exome : true
        const genomeCoverage = includeGenomeCoverage ? data.transcript.coverage.genome : true
        return exomeCoverage || genomeCoverage
      }}
    >
      {({ data }: any) => {
        const exomeCoverage = includeExomeCoverage ? data.transcript.coverage.exome : null
        const genomeCoverage = includeGenomeCoverage ? data.transcript.coverage.genome : null

        const coverageConfig =
          datasetId === 'exac'
            ? coverageConfigClassic(exomeCoverage, genomeCoverage)
            : coverageConfigNew(exomeCoverage, genomeCoverage)

        return (
          // @ts-expect-error TS(2769) FIXME: No overload matches this call.
          <CoverageTrack
            coverageOverThresholds={[1, 5, 10, 15, 20, 25, 30, 50, 100]}
            datasets={coverageConfig}
            filenameForExport={() => `${transcriptId}_coverage`}
            height={190}
          />
        )
      }}
    </Query>
  )
}

TranscriptCoverageTrack.defaultProps = {
  includeExomeCoverage: true,
  includeGenomeCoverage: true,
}

export default TranscriptCoverageTrack
