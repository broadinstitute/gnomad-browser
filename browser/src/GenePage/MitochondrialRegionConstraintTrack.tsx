import React from 'react'
import { Exon } from '../TranscriptPage/TranscriptPage'
import ConstraintTrack, { regionsInExons, RegionAttributeList } from '../ConstraintTrack'

export type MitochondrialConstraintRegion = {
  start: number
  stop: number
  oe: number
  oe_upper: number
  oe_lower: number
}

type Props = {
  geneSymbol: string
  constraintRegions: MitochondrialConstraintRegion[] | null
  exons: Exon[]
}

const constraintColor = '#9b001f'

const Legend = () => (
  <>
    <span>Regionally constrained interval</span>
    <svg width={50} height={25}>
      <rect x={10} y={3} width={30} height={10} stroke="#000" fill={constraintColor} />
    </svg>
  </>
)

type TooltipProps = {
  region: MitochondrialConstraintRegion
}

const Tooltip = ({ region, oeLabel }: TooltipProps & { oeLabel: string }) => {
  return (
    <RegionAttributeList>
      <div>
        <dt>Coordinates:</dt>
        <dd>{`M:${region.start}-${region.stop}`}</dd>
      </div>
      <div>
        <dt>{oeLabel}</dt>
        <dd>
          {region.oe.toFixed(3)} ({region.oe_lower.toFixed(3)}-{region.oe_upper.toFixed(3)})
        </dd>
      </div>
    </RegionAttributeList>
  )
}

const RNAGeneTooltip = ({ region }: TooltipProps) => {
  return Tooltip({ region, oeLabel: 'Observed/expected' })
}

const proteinGeneTooltip = ({ region }: TooltipProps) => {
  return Tooltip({ region, oeLabel: 'Missense observed/expected' })
}

const formattedOE = (region: MitochondrialConstraintRegion) => region.oe.toFixed(3)

const MitochondrialConstraintRegionTrack = ({ geneSymbol, constraintRegions, exons }: Props) => {
  if (constraintRegions === null) {
    return null
  }

  const isRNAGene = geneSymbol.startsWith('MT-R')
  const trackTitle = isRNAGene ? 'Regional constraint' : 'Regional missense constraint'

  return (
    <ConstraintTrack
      trackTitle={trackTitle}
      infobuttonTopic="TK-mitochondrial-gene-constraint"
      legend={<Legend />}
      valueFn={formattedOE}
      colorFn={() => constraintColor}
      tooltipComponent={isRNAGene ? RNAGeneTooltip : proteinGeneTooltip}
      allRegions={null}
      constrainedRegions={regionsInExons(constraintRegions, exons)}
    />
  )
}

export default MitochondrialConstraintRegionTrack
