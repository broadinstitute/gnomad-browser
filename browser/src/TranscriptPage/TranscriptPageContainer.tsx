import React from 'react'
import { Redirect } from 'react-router-dom'

import { DatasetId, referenceGenome } from '@gnomad/dataset-metadata/metadata'
import Query from '../Query'
import TranscriptPage from './TranscriptPage'

const operationName = 'Transcript'
const query = `
query ${operationName}($transcriptId: String!, $referenceGenome: ReferenceGenomeId!) {
  transcript(transcript_id: $transcriptId, reference_genome: $referenceGenome) {
    reference_genome
    transcript_id
    transcript_version
    chrom
    strand
    start
    stop
    exons {
      feature_type
      start
      stop
    }
    gnomad_constraint {
      exp_lof
      exp_mis
      exp_syn
      obs_lof
      obs_mis
      obs_syn
      oe_lof
      oe_lof_lower
      oe_lof_upper
      oe_mis
      oe_mis_lower
      oe_mis_upper
      oe_syn
      oe_syn_lower
      oe_syn_upper
      lof_z
      mis_z
      syn_z
      pLI
      flags
    }
    exac_constraint {
      exp_syn
      obs_syn
      syn_z
      exp_mis
      obs_mis
      mis_z
      exp_lof
      obs_lof
      lof_z
      pLI
    }
    gene {
      gene_id
      gene_version
      reference_genome
      symbol
      name
      canonical_transcript_id
      mane_select_transcript {
        ensembl_id
        ensembl_version
        refseq_id
        refseq_version
      }
      hgnc_id
      omim_id
      chrom
      start
      stop
      strand
      exons {
        feature_type
        start
        stop
      }
      flags
    }
  }
}
`

type Props = {
  datasetId: DatasetId
  transcriptId: string
}

const TranscriptPageContainer = ({ datasetId, transcriptId }: Props) => (
  <Query
    operationName={operationName}
    query={query}
    variables={{ transcriptId, referenceGenome: referenceGenome(datasetId) }}
    loadingMessage="Loading transcript"
    errorMessage="Unable to load transcript"
    success={(data: any) => data.transcript}
  >
    {({ data }: any) => {
      const { transcript } = data

      // Cannot query structural variants by transcript, redirect to gene page
      if (datasetId.startsWith('gnomad_sv')) {
        return <Redirect to={`/gene/${transcript.gene.gene_id}?dataset=${datasetId}`} />
      }

      return <TranscriptPage datasetId={datasetId} transcript={transcript} />
    }}
  </Query>
)

export default TranscriptPageContainer
