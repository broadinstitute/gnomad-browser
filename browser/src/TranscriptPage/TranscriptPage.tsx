import React, { useState } from 'react'
import styled from 'styled-components'

// @ts-expect-error TS(7016) FIXME: Could not find a declaration file for module '@gno... Remove this comment to see the full error message
import { TranscriptPlot } from '@gnomad/track-transcripts'

import ConstraintTable from '../ConstraintTable/ConstraintTable'
import {
  DatasetId,
  hasExomeCoverage,
  labelForDataset,
  ReferenceGenome,
} from '@gnomad/dataset-metadata/metadata'
import DocumentTitle from '../DocumentTitle'
import GeneFlags from '../GenePage/GeneFlags'
import GnomadPageHeading from '../GnomadPageHeading'
import InfoButton from '../help/InfoButton'
import RegionViewer from '../RegionViewer/ZoomableRegionViewer'
import { TrackPage, TrackPageSection } from '../TrackPage'
import { useWindowSize } from '../windowSize'

import MitochondrialTranscriptCoverageTrack from './MitochondrialTranscriptCoverageTrack'
import MitochondrialVariantsInTranscript from './MitochondrialVariantsInTranscript'
import TranscriptCoverageTrack from './TranscriptCoverageTrack'
import TranscriptInfo from './TranscriptInfo'
import TranscriptTrack from './TranscriptTrack'
import VariantsInTranscript from './VariantsInTranscript'

const TranscriptInfoColumnWrapper = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  margin-bottom: 1em;

  @media (max-width: 1200px) {
    flex-direction: column;
    align-items: center;
  }

  /* Matches responsive styles in AttributeList */
  @media (max-width: 600px) {
    align-items: stretch;
  }
`

const ControlPanel = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: flex-end;
  align-items: center;
  width: ${(props: any) => props.width}px;
  margin-left: ${(props: any) => props.marginLeft}px;

  @media (max-width: 600px) {
    flex-direction: column;
    align-items: flex-start;
    margin-left: 0;
  }
`

const Legend = styled.ul`
  display: flex;
  flex-direction: row;
  padding: 0;
  margin: 0.5em 0;
  list-style-type: none;
`

const LegendItemWrapper = styled.li`
  display: flex;
  align-items: stretch;
  margin-left: 1em;
`

const Label = styled.label`
  user-select: none;
  display: flex;
  flex-direction: row;
  align-items: center;
`

const CheckboxInput = styled.input.attrs({ type: 'checkbox' })`
  margin-right: 0.5em;
`

const LegendSwatch = styled.span`
  display: flex;
  align-items: center;
  width: 16px;
  margin-left: 0.5em;

  &::before {
    content: '';
    display: inline-block;
    width: 16px;
    height: ${(props: any) => props.height}px;
    background: ${(props: any) => props.color};
  }
`

const transcriptFeatureAttributes = {
  exon: {
    fill: '#bdbdbd',
    height: 4,
  },
  CDS: {
    fill: '#424242',
    height: 10,
  },
  UTR: {
    fill: '#424242',
    height: 4,
  },
}

type Strand = '+' | '-'

export type Transcript = {
  transcript_id: string
  transcript_version: string
  reference_genome: ReferenceGenome
  chrom: string
  start: number
  stop: number
  strand: Strand
  exons: {
    feature_type: string
    start: number
    stop: number
  }[]
  gene: Gene
}

export type Gene = {
  gene_id: string
  gene_version: string
  reference_genome: ReferenceGenome
  symbol: string
  name?: string
  strand: Strand
  exons: {
    feature_type: string
    start: number
    stop: number
  }[]
  canonical_transcript_id?: string
  mane_select_transcript?: {
    ensembl_id: string
    ensembl_version: string
    refseq_id: string
    refseq_version: string
  }
  flags: string[]
}

type Props = {
  datasetId: DatasetId
  transcript: Transcript
}

const TranscriptPage = ({ datasetId, transcript }: Props) => {
  const [includeUTRs, setIncludeUTRs] = useState(false)

  const { width: windowWidth } = useWindowSize()
  const isSmallScreen = windowWidth < 900

  // Subtract 30px for padding on Page component
  const regionViewerWidth = windowWidth - 30

  const cdsCompositeExons = transcript.exons.filter((exon) => exon.feature_type === 'CDS')
  const hasCodingExons = cdsCompositeExons.length > 0
  const hasUTRs = transcript.exons.some((exon) => exon.feature_type === 'UTR')

  const isCodingTranscript = transcript.exons.some((exon) => exon.feature_type === 'CDS')

  const regionViewerRegions = transcript.exons
    .filter(
      (exon) =>
        exon.feature_type === 'CDS' ||
        (exon.feature_type === 'UTR' && includeUTRs) ||
        (exon.feature_type === 'exon' && !isCodingTranscript)
    )
    .map((exon) => ({
      start: Math.max(1, exon.start - 75),
      stop: exon.stop + 75,
    }))

  const [zoomRegion, setZoomRegion] = useState(null)

  return (
    <TrackPage>
      <TrackPageSection>
        <DocumentTitle title={`${transcript.transcript_id} | ${labelForDataset(datasetId)}`} />
        <GnomadPageHeading
          selectedDataset={datasetId}
          datasetOptions={{
            includeShortVariants: true,
            includeStructuralVariants: false,
            includeExac: transcript.chrom !== 'M',
            includeGnomad2: transcript.chrom !== 'M',
            includeGnomad3: true,
            includeGnomad3Subsets: transcript.chrom !== 'M',
          }}
        >
          Transcript: {transcript.transcript_id}.{transcript.transcript_version}
        </GnomadPageHeading>
        <TranscriptInfoColumnWrapper>
          <div style={{ maxWidth: '50%' }}>
            <TranscriptInfo transcript={transcript} />
            <GeneFlags gene={transcript.gene} />
          </div>
          <div>
            <h2>Constraint {transcript.chrom !== 'M' && <InfoButton topic="constraint" />}</h2>
            <ConstraintTable datasetId={datasetId} geneOrTranscript={transcript} />
          </div>
        </TranscriptInfoColumnWrapper>
      </TrackPageSection>
      <RegionViewer
        contextType="transcript"
        leftPanelWidth={115}
        width={regionViewerWidth}
        regions={regionViewerRegions}
        rightPanelWidth={isSmallScreen ? 0 : 80}
        renderOverview={({ scalePosition, width: overviewWidth }: any) => (
          <TranscriptPlot
            height={10}
            scalePosition={scalePosition}
            showNonCodingExons
            showUTRs={includeUTRs}
            transcript={transcript}
            width={overviewWidth}
          />
        )}
        zoomRegion={zoomRegion}
        onChangeZoomRegion={setZoomRegion}
      >
        {transcript.chrom === 'M' ? (
          <MitochondrialTranscriptCoverageTrack
            datasetId={datasetId}
            transcriptId={transcript.transcript_id}
          />
        ) : (
          <TranscriptCoverageTrack
            datasetId={datasetId}
            transcriptId={transcript.transcript_id}
            includeExomeCoverage={hasExomeCoverage(datasetId)}
          />
        )}

        {/* @ts-expect-error TS(2769) FIXME: No overload matches this call. */}
        <ControlPanel marginLeft={100} width={regionViewerWidth - 100 - (isSmallScreen ? 0 : 80)}>
          Include:
          <Legend>
            <LegendItemWrapper>
              <Label htmlFor="include-cds-regions">
                <CheckboxInput
                  checked={hasCodingExons}
                  disabled
                  id="include-cds-regions"
                  onChange={() => {}}
                />
                Coding regions (CDS)
                <LegendSwatch
                  color={transcriptFeatureAttributes.CDS.fill}
                  // @ts-expect-error TS(2769) FIXME: No overload matches this call.
                  height={transcriptFeatureAttributes.CDS.height}
                />
              </Label>
            </LegendItemWrapper>

            <LegendItemWrapper>
              <Label htmlFor="include-utr-regions">
                <CheckboxInput
                  checked={includeUTRs}
                  disabled={!hasUTRs}
                  id="include-utr-regions"
                  onChange={(e: any) => {
                    setIncludeUTRs(e.target.checked)
                  }}
                />
                Untranslated regions (UTRs)
                <LegendSwatch
                  color={transcriptFeatureAttributes.UTR.fill}
                  // @ts-expect-error TS(2769) FIXME: No overload matches this call.
                  height={transcriptFeatureAttributes.UTR.height}
                />
              </Label>
            </LegendItemWrapper>

            <LegendItemWrapper>
              <Label htmlFor="include-nc-regions">
                <CheckboxInput
                  checked={!isCodingTranscript}
                  disabled
                  id="include-nc-regions"
                  onChange={() => {}}
                />
                Non-coding regions
                <LegendSwatch
                  color={transcriptFeatureAttributes.exon.fill}
                  // @ts-expect-error TS(2769) FIXME: No overload matches this call.
                  height={transcriptFeatureAttributes.exon.height}
                />
              </Label>
            </LegendItemWrapper>
          </Legend>
        </ControlPanel>

        <div style={{ margin: '1em 0' }}>
          <TranscriptTrack transcript={transcript} showUTRs={includeUTRs} />
        </div>

        {transcript.chrom === 'M' ? (
          <MitochondrialVariantsInTranscript
            datasetId={datasetId}
            transcript={transcript}
            zoomRegion={zoomRegion}
          />
        ) : (
          <VariantsInTranscript
            datasetId={datasetId}
            // @ts-expect-error TS(2322) FIXME: Type '{ datasetId: string; includeUTRs: boolean; t... Remove this comment to see the full error message
            includeUTRs={includeUTRs}
            transcript={transcript}
            zoomRegion={zoomRegion}
          />
        )}
      </RegionViewer>
    </TrackPage>
  )
}

export default TranscriptPage
