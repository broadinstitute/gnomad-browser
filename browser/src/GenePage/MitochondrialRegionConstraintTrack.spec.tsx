import React from 'react'
import renderer from 'react-test-renderer'
import { expect, test } from '@jest/globals'
import MitochondrialRegionConstraintTrack, {
  MitochondrialConstraintRegion,
} from './MitochondrialRegionConstraintTrack'
import { Exon } from '../TranscriptPage/TranscriptPage'
// @ts-expect-error
import { RegionViewerContext } from '@gnomad/region-viewer'

const childProps = {
  centerPanelWidth: 3,
  isPositionDefined: true,
  leftPanelWidth: 4,
  regions: [],
  rightPanelWidth: 5,
  scalePosition: (i: number) => i,
}

const exons: Exon[] = [{ feature_type: 'CDS', start: 123, stop: 1999 }]
const constraintRegions: MitochondrialConstraintRegion[] = [
  { start: 555, stop: 666, oe: 0.45, oe_lower: 0.37, oe_upper: 0.47 },
  { start: 777, stop: 888, oe: 0.56, oe_lower: 0.52, oe_upper: 0.59 },
]

describe('MitochondrialRegionConstraintTrack for a protein gene', () => {
  test('track has no unexpected changes when gene has constraint', () => {
    const tree = renderer.create(
      <RegionViewerContext.Provider value={childProps}>
        <MitochondrialRegionConstraintTrack
          geneSymbol="MT-ATP6"
          constraintRegions={constraintRegions}
          exons={exons}
        />
      </RegionViewerContext.Provider>
    )
    expect(tree).toMatchSnapshot()
  })

  test('track has no unexpected changes when no constraint for gene', () => {
    const tree = renderer.create(
      <RegionViewerContext.Provider value={childProps}>
        <MitochondrialRegionConstraintTrack
          geneSymbol="MT-ATP6"
          constraintRegions={null}
          exons={exons}
        />
      </RegionViewerContext.Provider>
    )
    expect(tree).toMatchSnapshot()
  })
})

describe('MitochondrialRegionConstraintTrack for an RNA gene', () => {
  test('track has no unexpected changes when gene has constraint', () => {
    const tree = renderer.create(
      <RegionViewerContext.Provider value={childProps}>
        <MitochondrialRegionConstraintTrack
          geneSymbol="MT-RNR1"
          constraintRegions={constraintRegions}
          exons={exons}
        />
      </RegionViewerContext.Provider>
    )
    expect(tree).toMatchSnapshot()
  })

  test('track has no unexpected changes when no constraint for gene', () => {
    const tree = renderer.create(
      <RegionViewerContext.Provider value={childProps}>
        <MitochondrialRegionConstraintTrack
          geneSymbol="MT-RNR1"
          constraintRegions={null}
          exons={exons}
        />
      </RegionViewerContext.Provider>
    )
    expect(tree).toMatchSnapshot()
  })
})
