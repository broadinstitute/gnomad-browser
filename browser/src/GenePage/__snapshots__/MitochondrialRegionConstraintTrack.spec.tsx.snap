// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MitochondrialRegionConstraintTrack for a protein gene track has no unexpected changes when gene has constraint 1`] = `
<div
  className="ConstraintTrack__Wrapper-sc-113yvfg-1 iBhwwK"
>
  <div
    className="Track__OuterWrapper-sc-1sdyh2h-0 bBMGlf"
  >
    <div
      className="Track__InnerWrapper-sc-1sdyh2h-1 cEOGGC"
    >
      <div
        className="Track__SidePanel-sc-1sdyh2h-3 iSYzDq"
        style={
          {
            "width": 4,
          }
        }
      >
        <div
          className="ConstraintTrack__SidePanel-sc-113yvfg-2 dJMFds"
        >
          <span>
            Regional missense constraint
          </span>
          <button
            className="InfoButton__Button-sc-13t5e82-0 gxsoyZ"
            onClick={[Function]}
            type="button"
          >
            <img
              alt=""
              aria-hidden="true"
              src="test-file-stub"
            />
            <span
              style={
                {
                  "border": "0",
                  "clip": "rect(0 0 0 0)",
                  "height": "1px",
                  "margin": "-1px",
                  "overflow": "hidden",
                  "padding": "0",
                  "position": "absolute",
                  "whiteSpace": "nowrap",
                  "width": "1px",
                }
              }
            >
              More information
            </span>
          </button>
        </div>
      </div>
      <div
        className="Track__CenterPanel-sc-1sdyh2h-4 iAyrrk"
        style={
          {
            "width": 3,
          }
        }
      >
        <div
          className="ConstraintTrack__TopPanel-sc-113yvfg-3 imJcbb"
        >
          <div
            className="ConstraintTrack__LegendWrapper-sc-113yvfg-4 kkHtue"
          >
            <span>
              Regionally constrained interval
            </span>
            <svg
              height={25}
              width={50}
            >
              <rect
                fill="#9b001f"
                height={10}
                stroke="#000"
                width={30}
                x={10}
                y={3}
              />
            </svg>
          </div>
        </div>
        <div
          className="ConstraintTrack__PlotWrapper-sc-113yvfg-0 fHKmMU"
        >
          <svg
            height={55}
            width={3}
          >
            <rect
              height={1}
              width={3}
              x={0}
              y={7.5}
            />
            <g
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              <rect
                fill="#9b001f"
                height={15}
                stroke="black"
                width={111}
                x={555}
                y={1}
              />
            </g>
            <g
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              <rect
                fill="#9b001f"
                height={15}
                stroke="black"
                width={111}
                x={777}
                y={1}
              />
            </g>
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`MitochondrialRegionConstraintTrack for a protein gene track has no unexpected changes when no constraint for gene 1`] = `null`;

exports[`MitochondrialRegionConstraintTrack for an RNA gene track has no unexpected changes when gene has constraint 1`] = `
<div
  className="ConstraintTrack__Wrapper-sc-113yvfg-1 iBhwwK"
>
  <div
    className="Track__OuterWrapper-sc-1sdyh2h-0 bBMGlf"
  >
    <div
      className="Track__InnerWrapper-sc-1sdyh2h-1 cEOGGC"
    >
      <div
        className="Track__SidePanel-sc-1sdyh2h-3 iSYzDq"
        style={
          {
            "width": 4,
          }
        }
      >
        <div
          className="ConstraintTrack__SidePanel-sc-113yvfg-2 dJMFds"
        >
          <span>
            Regional constraint
          </span>
          <button
            className="InfoButton__Button-sc-13t5e82-0 gxsoyZ"
            onClick={[Function]}
            type="button"
          >
            <img
              alt=""
              aria-hidden="true"
              src="test-file-stub"
            />
            <span
              style={
                {
                  "border": "0",
                  "clip": "rect(0 0 0 0)",
                  "height": "1px",
                  "margin": "-1px",
                  "overflow": "hidden",
                  "padding": "0",
                  "position": "absolute",
                  "whiteSpace": "nowrap",
                  "width": "1px",
                }
              }
            >
              More information
            </span>
          </button>
        </div>
      </div>
      <div
        className="Track__CenterPanel-sc-1sdyh2h-4 iAyrrk"
        style={
          {
            "width": 3,
          }
        }
      >
        <div
          className="ConstraintTrack__TopPanel-sc-113yvfg-3 imJcbb"
        >
          <div
            className="ConstraintTrack__LegendWrapper-sc-113yvfg-4 kkHtue"
          >
            <span>
              Regionally constrained interval
            </span>
            <svg
              height={25}
              width={50}
            >
              <rect
                fill="#9b001f"
                height={10}
                stroke="#000"
                width={30}
                x={10}
                y={3}
              />
            </svg>
          </div>
        </div>
        <div
          className="ConstraintTrack__PlotWrapper-sc-113yvfg-0 fHKmMU"
        >
          <svg
            height={55}
            width={3}
          >
            <rect
              height={1}
              width={3}
              x={0}
              y={7.5}
            />
            <g
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              <rect
                fill="#9b001f"
                height={15}
                stroke="black"
                width={111}
                x={555}
                y={1}
              />
            </g>
            <g
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              <rect
                fill="#9b001f"
                height={15}
                stroke="black"
                width={111}
                x={777}
                y={1}
              />
            </g>
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`MitochondrialRegionConstraintTrack for an RNA gene track has no unexpected changes when no constraint for gene 1`] = `null`;
